python/TestHarness/util.py

```diff
@@ -354,31 +354,71 @@ def getSlepcVersion(libmesh_dir):
 
     return major_version.pop() + '.' + minor_version.pop() + '.' + subminor_version.pop()
 
+def checkLogicVersionSplits(target, splits, logic_and, package):
+    status, logic_reason, version = [],[],[]
+    for split in splits:
+        (lstatus, lreason, lversion) = checkLogicVersionSingle(target, split, package)
+        status.append(lstatus)
+        logic_reason.append(lreason)
+        version.append(lversion)
+
+    if logic_and:
+        if status[0] and status[1]:
+            return (True, None, version[0])
+        else:
+            return (False, '&&', version[0]+' '+version[1])
+    else:
+        if status[0] or status[1]:
+            return (True, None, version[0])
+        else:
+            return (True, '||', version[0]+ ' '+version[1])
+
+    return (False, None, None)
+
+def checkLogicVersionSingle(checks, iversion, package):
+    logic, version = re.search(r'(.*?)(\d\S+)', iversion).groups()
+    if logic == '' or logic == '=':
+        if version == checks[package]:
+            return (True, None, version)
+        else:
+            return (False, '!=', version)
+
+    # Logical match
+    if logic == '>' and checks[package][0:5] > version[0:5]:
+        return (True, None, version)
+    elif logic == '>=' and checks[package][0:5] >= version[0:5]:
+        return (True, None, version)
+    elif logic == '<' and checks[package][0:5] < version[0:5]:
+        return (True, None, version)
+    elif logic == '<=' and checks[package][0:5] <= version[0:5]:
+        return (True, None, version)
+
+    return (False, logic, version)
+
+def checkVersion(checks, test, package):
+    for version in test[package]:
+        logic_and = False
+        if version.find("&&") != -1:
+            logic_and = True
+        splits = re.split('[(&&) | (||)]+', version)
+        if len(splits) == 1:
+            return checkLogicVersionSingle(checks, version, package)
+        elif len(splits) == 2:
+            return checkLogicVersionSplits(checks, splits, logic_and, package)
```

Comment:
This all works, but isn't the cleanest way to do things. "re" can be used to do the split and give back the captured token. I think all of this logic could be compressed into a few lines (maybe 10)... I guess it's not worth really cleaning it up, but keep in mind this is Python, not C.